### ISSUE TITLE
isaac_ros_dope

### ISSUE THREAD

**Issue by czy1998916** [NVlabs/Deep_Object_Pose#410]
Dear author, hello. I am using isaac_ros-d ope to infer objects of interest. Here are the commands I have run and the terminal output
`rx@ubuntu:~/workspaces/isaac_ros-dev$ ros2 launch isaac_ros_dope isaac_ros_dope_tensor_rt.launch.py 
[INFO] [launch]: All log files can be found below /home/rx/.ros/log/2025-06-19-19-25-29-139862-ubuntu-90482
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [component_container_mt-1]: process started with pid [90495]
[component_container_mt-1] [INFO] [1750332329.571044796] [dope_container]: Load Library: /opt/ros/humble/lib/libtensor_rt_node.so
[component_container_mt-1] [INFO] [1750332329.677057465] [dope_container]: Found class: rclcpp_components::NodeFactoryTemplate<nvidia::isaac_ros::dnn_inference::TensorRTNode>
[component_container_mt-1] [INFO] [1750332329.677168750] [dope_container]: Instantiate class: rclcpp_components::NodeFactoryTemplate<nvidia::isaac_ros::dnn_inference::TensorRTNode>
[component_container_mt-1] [INFO] [1750332329.695735059] [dope_inference]: [NitrosNode] Initializing NitrosNode
[component_container_mt-1] [INFO] [1750332329.696182533] [NitrosContext]: [NitrosContext] Loading extension: gxf/lib/std/libgxf_std.so
[component_container_mt-1] [INFO] [1750332329.697730792] [NitrosContext]: [NitrosContext] Loading extension: gxf/lib/libgxf_isaac_gxf_helpers.so
[component_container_mt-1] [INFO] [1750332329.702866559] [NitrosContext]: [NitrosContext] Loading extension: gxf/lib/libgxf_isaac_sight.so
[component_container_mt-1] [INFO] [1750332329.711026851] [NitrosContext]: [NitrosContext] Loading extension: gxf/lib/libgxf_isaac_atlas.so
[component_container_mt-1] 2025-06-19 19:25:29.723 WARN  gxf/std/program.cpp@538: No GXF scheduler specified.
[component_container_mt-1] [INFO] [1750332329.724171022] [dope_inference]: [TensorRTNode] Set input data format to: "nitros_tensor_list_nchw_rgb_f32"
[component_container_mt-1] [INFO] [1750332329.724234759] [dope_inference]: [TensorRTNode] Set output data format to: "nitros_tensor_list_nhwc_rgb_f32"
[component_container_mt-1] [INFO] [1750332329.724307552] [dope_inference]: [NitrosNode] Starting NitrosNode
[component_container_mt-1] [INFO] [1750332329.729402363] [dope_inference]: [NitrosNode] Loading extensions
[component_container_mt-1] [INFO] [1750332329.729662593] [dope_inference]: [NitrosContext] Loading extension: gxf/lib/multimedia/libgxf_multimedia.so
[component_container_mt-1] [INFO] [1750332329.733327725] [dope_inference]: [NitrosContext] Loading extension: gxf/lib/libgxf_isaac_message_compositor.so
[component_container_mt-1] [INFO] [1750332329.735250154] [dope_inference]: [NitrosContext] Loading extension: gxf/lib/cuda/libgxf_cuda.so
[component_container_mt-1] [INFO] [1750332329.739617935] [dope_inference]: [NitrosContext] Loading extension: gxf/lib/serialization/libgxf_serialization.so
[component_container_mt-1] [INFO] [1750332329.741420856] [dope_inference]: [NitrosContext] Loading extension: gxf/lib/libgxf_isaac_tensor_rt.so
[component_container_mt-1] [INFO] [1750332329.742763344] [dope_inference]: [NitrosNode] Loading graph to the optimizer
[component_container_mt-1] [INFO] [1750332329.745095587] [dope_inference]: [NitrosNode] Running optimization
[component_container_mt-1] [INFO] [1750332329.768512412] [dope_inference]: [NitrosNode] Obtaining graph IO group info from the optimizer
[component_container_mt-1] [INFO] [1750332329.770523632] [dope_inference]: [NitrosPublisherSubscriberGroup] Pinning the component "inference/rx" (type="nvidia::gxf::DoubleBufferReceiver") to use its compatible format only: "nitros_tensor_list_nchw_rgb_f32"
[component_container_mt-1] [INFO] [1750332329.770661794] [dope_inference]: [NitrosPublisherSubscriberGroup] Pinning the component "sink/sink" (type="nvidia::isaac_ros::MessageRelay") to use its compatible format only: "nitros_tensor_list_nhwc_rgb_f32"
[component_container_mt-1] [INFO] [1750332329.777734772] [dope_inference]: [NitrosNode] Starting negotiation...
[INFO] [launch_ros.actions.load_composable_nodes]: Loaded node '/dope_inference' in container '/dope_container'
[component_container_mt-1] [INFO] [1750332329.779382253] [dope_container]: Load Library: /opt/ros/humble/lib/libresize_node.so
.....................
[component_container_mt-1] [INFO] [1750332842.202532611] [dope_encoder.reshape_node]: [NitrosSubscriber] Use the negotiated data format: "nitros_tensor_list_nchw_rgb_f32"
[component_container_mt-1] [INFO] [1750332842.202799541] [dope_encoder.reshape_node]: [NitrosNode] Exporting the final graph based on the negotiation results
[component_container_mt-1] [INFO] [1750332842.211550859] [dope_encoder.reshape_node]: [NitrosNode] Wrote the final top level YAML graph to "/tmp/isaac_ros_nitros/graphs/XWAWYZZSMK/XWAWYZZSMK.yaml"
[component_container_mt-1] [INFO] [1750332842.211661702] [dope_encoder.reshape_node]: [NitrosNode] Loading application
[component_container_mt-1] [INFO] [1750332842.213780855] [dope_encoder.reshape_node]: In ReshapeNode postLoadGraphCallback().
[component_container_mt-1] [INFO] [1750332842.213891697] [dope_encoder.reshape_node]: [NitrosNode] Initializing and running GXF graph
[component_container_mt-1] [INFO] [1750332842.352944614] [dope_encoder.reshape_node]: [NitrosNode] Node was started`
Here is the content of my launch file
`import os

from ament_index_python.packages import get_package_share_directory
import launch
from launch.actions import DeclareLaunchArgument, IncludeLaunchDescription
from launch.launch_description_sources import PythonLaunchDescriptionSource
from launch.substitutions import LaunchConfiguration
from launch_ros.actions import ComposableNodeContainer
from launch_ros.descriptions import ComposableNode

def generate_launch_description():
    """Generate launch description for DOPE encoder->TensorRT->DOPE decoder."""
    DEFAULT_MODEL_FILE_NAME = 'dope_ketchup_pol.onnx'
    default_model_file_path = os.path.dirname(os.path.abspath(
        __file__)) + '/../../test/models/' + DEFAULT_MODEL_FILE_NAME
    launch_args = [
        DeclareLaunchArgument(
            'input_image_width',
            default_value='1280',
            description='The input image width'),
        DeclareLaunchArgument(
            'input_image_height',
            default_value='720',
            description='The input image height'),
        DeclareLaunchArgument(
            'network_image_width',
            default_value='1280',
            description='The input image width that the network expects'),
        DeclareLaunchArgument(
            'network_image_height',
            default_value='720',
            description='The input image height that the network expects'),
        DeclareLaunchArgument(
            'encoder_image_mean',
            default_value='[0.485, 0.456, 0.406]',
            description='The mean for image normalization'),
        DeclareLaunchArgument(
            'encoder_image_stddev',
            default_value='[0.229, 0.224, 0.225]',
            description='The standard deviation for image normalization'),
        DeclareLaunchArgument(
            'model_file_path',
            default_value='/home/rx/workspaces/isaac_ros-dev/isaac_ros_assets/models/dope/vita.onnx',
            description='The absolute file path to the ONNX file'),
        DeclareLaunchArgument(
            'engine_file_path',
            default_value='/tmp/trt_engine.plan',
            description='The absolute file path to the TensorRT engine file'),
        DeclareLaunchArgument(
            'input_tensor_names',
            default_value='["input_tensor"]',
            description='A list of tensor names to bound to the specified input binding names'),
        DeclareLaunchArgument(
            'input_binding_names',
            default_value='["input"]',
            description='A list of input tensor binding names (specified by model)'),
        DeclareLaunchArgument(
            'input_tensor_formats',
            default_value='["nitros_tensor_list_nchw_rgb_f32"]',
            description='The nitros format of the input tensors'),
        DeclareLaunchArgument(
            'output_tensor_names',
            default_value='["output"]',
            description='A list of tensor names to bound to the specified output binding names'),
        DeclareLaunchArgument(
            'output_binding_names',
            default_value='["output"]',
            description='A  list of output tensor binding names (specified by model)'),
        DeclareLaunchArgument(
            'output_tensor_formats',
            default_value='["nitros_tensor_list_nhwc_rgb_f32"]',
            description='The nitros format of the output tensors'),
        DeclareLaunchArgument(
            'tensorrt_verbose',
            default_value='False',
            description='Whether TensorRT should verbosely log or not'),
        DeclareLaunchArgument(
            'object_name',
            default_value='vita',
            description='The object class that the DOPE network is detecting'),
        DeclareLaunchArgument(
            'force_engine_update',
            default_value='False',
            description='Whether TensorRT should update the TensorRT engine file or not'),
        DeclareLaunchArgument(
            'map_peak_threshold',
            default_value='0.1',
            description='The minimum value of a peak in a DOPE belief map'),
        DeclareLaunchArgument(
            'enable_tf_publishing',
            default_value='False',
            description='Whether Dope Decoder will broadcast poses to the TF tree or not')
    ]

    # DNN Image Encoder parameters
    input_image_width = LaunchConfiguration('input_image_width')
    input_image_height = LaunchConfiguration('input_image_height')
    network_image_width = LaunchConfiguration('network_image_width')
    network_image_height = LaunchConfiguration('network_image_height')
    encoder_image_mean = LaunchConfiguration('encoder_image_mean')
    encoder_image_stddev = LaunchConfiguration('encoder_image_stddev')

    # Tensor RT parameters
    model_file_path = LaunchConfiguration('model_file_path')
    engine_file_path = LaunchConfiguration('engine_file_path')
    input_tensor_names = LaunchConfiguration('input_tensor_names')
    input_binding_names = LaunchConfiguration('input_binding_names')
    input_tensor_formats = LaunchConfiguration('input_tensor_formats')
    output_tensor_names = LaunchConfiguration('output_tensor_names')
    output_binding_names = LaunchConfiguration('output_binding_names')
    output_tensor_formats = LaunchConfiguration('output_tensor_formats')
    tensorrt_verbose = LaunchConfiguration('tensorrt_verbose')
    force_engine_update = LaunchConfiguration('force_engine_update')

    # DOPE Decoder parameters
    object_name = LaunchConfiguration('object_name')
    enable_tf_publishing = LaunchConfiguration('enable_tf_publishing')
    map_peak_threshold = LaunchConfiguration('map_peak_threshold')

    encoder_dir = get_package_share_directory('isaac_ros_dnn_image_encoder')
    dope_encoder_launch = IncludeLaunchDescription(
        PythonLaunchDescriptionSource(
            [os.path.join(encoder_dir, 'launch', 'dnn_image_encoder.launch.py')]
        ),
        launch_arguments={
            'input_image_width': input_image_width,
            'input_image_height': input_image_height,
            'network_image_width': network_image_width,
            'network_image_height': network_image_height,
            'image_mean': encoder_image_mean,
            'image_stddev': encoder_image_stddev,
            'attach_to_shared_component_container': 'True',
            'component_container_name': 'dope_container',
            'dnn_image_encoder_namespace': 'dope_encoder',
            'image_input_topic': '/camera/camera/color/image_raw',
            'camera_info_input_topic': '/camera/camera/color/camera_info',
            'tensor_output_topic': '/tensor_pub',
            'keep_aspect_ratio': 'False'
        }.items(),
    )

    dope_inference_node = ComposableNode(
        name='dope_inference',
        package='isaac_ros_tensor_rt',
        plugin='nvidia::isaac_ros::dnn_inference::TensorRTNode',
        parameters=[{
            'model_file_path': model_file_path,
            'engine_file_path': engine_file_path,
            'input_tensor_names': input_tensor_names,
            'input_binding_names': input_binding_names,
            'input_tensor_formats': input_tensor_formats,
            'output_tensor_names': output_tensor_names,
            'output_binding_names': output_binding_names,
            'output_tensor_formats': output_tensor_formats,
            'verbose': tensorrt_verbose,
            'force_engine_update': force_engine_update
        }])

    dope_decoder_node = ComposableNode(
        name='dope_decoder',
        package='isaac_ros_dope',
        plugin='nvidia::isaac_ros::dope::DopeDecoderNode',
        parameters=[{
            'object_name': object_name,
            'enable_tf_publishing': enable_tf_publishing,
            'map_peak_threshold': map_peak_threshold,
        }],
        remappings=[('belief_map_array', 'tensor_sub'),
                    ('dope/detections', 'detections'),
                    ('camera_info', '/dope_encoder/crop/camera_info')]
    )

    container = ComposableNodeContainer(
        name='dope_container',
        namespace='',
        package='rclcpp_components',
        executable='component_container_mt',
        composable_node_descriptions=[dope_inference_node, dope_decoder_node],
        output='screen',
        arguments=['--ros-args', '--log-level', 'info']
    )

    final_launch_description = launch_args + [container, dope_encoder_launch]
    return launch.LaunchDescription(final_launch_description)`
But/detections have no output

![Image](https://github.com/user-attachments/assets/c1a30d4d-1e48-485b-88fd-27cd1a3781ca)
`header:
  stamp:
    sec: 1750330287
    nanosec: 268315430
  frame_id: camera_color_optical_frame
detections: []
---`
My vita. pth model is fine

![Image](https://github.com/user-attachments/assets/aadb0bb4-3af0-451c-a376-982c449e7122)
Excuse me, can someone tell me what the problem is

**Comment by nv-jeff** [comment]
The authors of DOPE do not work on or maintain the Isaac ROS DOPE implementation; you should ask your question there (https://github.com/NVIDIA-ISAAC-ROS/isaac_ros_pose_estimation).

However, I would recommend stepping back from the Isaac ROS code until you are certain your model is trained well. Evaluate the accuracy of your trained model using the DOPE inference code outside of the Isaac ROS node. A well trained model should be able to acheive 70% or greater accuracy, but don't expect 100% -- especially if your training data domain differs significantly from your inference test images.

Some of the inference parameters in `config_pose.yaml` (e.g. `thresh_map`) can be changed in order to allow more, but less accurate, detections.  These parameters are not documented, but looking through the code and reading the paper can help you understand them
